lib/modules/manager/npm/extract/npm.ts

```diff
@@ -8,15 +9,39 @@ export async function getNpmLock(filePath: string): Promise<LockFile> {
   try {
     const lockParsed = JSON.parse(lockRaw);
     const lockedVersions: Record<string, string> = {};
-    for (const [entry, val] of Object.entries(
-      (lockParsed.dependencies || {}) as LockFileEntry
-    )) {
+    const { packages, lockfileVersion } = extractPackages(lockParsed);
+    for (const [entry, val] of Object.entries(packages)) {
       logger.trace({ entry, version: val.version });
       lockedVersions[entry] = val.version;
     }
-    return { lockedVersions, lockfileVersion: lockParsed.lockfileVersion };
+    return { lockedVersions, lockfileVersion };
   } catch (err) {
     logger.debug({ filePath, err }, 'Warning: Exception parsing npm lock file');
     return { lockedVersions: {} };
   }
 }
+
+export function extractPackages(lockParsed: any): {
+  packages: LockFileEntry;
+  lockfileVersion: number;
+} {
+  const packageLockPreV3ParseResult =
+    PackageLockPreV3Schema.safeParse(lockParsed);
```

Comment:
we should do this with schema transforms instead. so first do a basic schema which only has lockfile version, then use transform to call the appropriate specific schema with known lockfile version. 

@zharinov can you do a suggestion?

Reply:
@zharinov did you see the comment above? a suggestion would be much appreciated.

Reply:
Hello, please merge this one into your branch: https://github.com/mjunker/renovate/pull/2